### PR TITLE
Improve documentation of Subscriber::record and Subscriber::event

### PR DESCRIPTION
**Pull request by MSleepyPanda**
The example code is ignored, because the macros would require pulling in `tokio_trace` as a dependency for `tokio_trace_core`.

Closes https://github.com/tokio-rs/tokio/issues/1007

### CODE REVIEW

**Comment by MSleepyPanda** [comment]
cc @hawkw

File: tokio-trace/tokio-trace-core/src/subscriber.rs [hunk 1]
```diff
@@ -158,9 +158,30 @@ pub trait Subscriber: 'static {
 
     /// Record a set of values on a span.
     ///
+    /// This method will be invoked when value is recorded on a span.
+    /// Recording multiple values for the same field is possible,
+    /// but the actual behaviour is implementation defined by the subscriber.
```
Comment (LucioFranco) [inline tokio-trace/tokio-trace-core/src/subscriber.rs]: I would reword this as `but the actual behavior is an implementation detail of the subscriber.`

File: tokio-trace/tokio-trace-core/src/subscriber.rs [hunk 2]
```diff
@@ -158,9 +158,30 @@ pub trait Subscriber: 'static {
...
+    ///
+    /// Keep in mind that a span might not provide values for every field it declares.
```
Comment (LucioFranco) [inline tokio-trace/tokio-trace-core/src/subscriber.rs]: TIOLI `keep in mind that a span may not provide a value for each fieled it declares.`

File: tokio-trace/tokio-trace-core/src/subscriber.rs [hunk 3]
```diff
@@ -187,6 +208,12 @@ pub trait Subscriber: 'static {
 
     /// Records that an [`Event`] has occurred.
     ///
+    /// This method will be invoked, when a `trace!`, `info!` or other event is constructed.
```
Comment (LucioFranco) [inline tokio-trace/tokio-trace-core/src/subscriber.rs]: TIOLI `or another event is constructed`

File: tokio-trace/tokio-trace-core/src/subscriber.rs [hunk 4]
```diff
@@ -187,6 +208,12 @@ pub trait Subscriber: 'static {
...
+    ///
+    /// The key difference between this method and `record` is that an event refers to the specific
```
Comment (hawkw) [inline tokio-trace/tokio-trace-core/src/subscriber.rs]: I would have just said here that the difference between this method and `record` is that `record` is called when a value is recorded for a field defined by a span, while `event` is called when a new event occurs. 

I think the semantic difference between recording a field on a span and recording an event ought to be discussed in the documentation for the `Event` type, not here. The docs on `Subscriber` could link to that.

File: tokio-trace/tokio-trace-core/src/subscriber.rs (hunk 2 above)
Comment (MSleepyPanda) [reply tokio-trace/tokio-trace-core/src/subscriber.rs]: I've settled in between, "may" not sounds like it is forbidden

File: tokio-trace/tokio-trace-core/src/subscriber.rs (hunk 4 above)
Comment (MSleepyPanda) [reply tokio-trace/tokio-trace-core/src/subscriber.rs]: This makes sense!

`Event` is already linked above, so i didn't include it as well

**Comment by MSleepyPanda** [comment]
Thank you for your feedback! I've updated the PR.

I'll rebase when this is good to go :)

File: tokio-trace/tokio-trace-core/src/subscriber.rs [hunk 5]
```diff
@@ -187,13 +217,22 @@ pub trait Subscriber: 'static {
 
     /// Records that an [`Event`] has occurred.
     ///
+    /// This method will be invoked when an Event is constructed by
+    /// the `Event`'s [`dispatch` method]. This happens internally
```
Comment (hawkw) [inline tokio-trace/tokio-trace-core/src/subscriber.rs]: Nit: I might add a 
```suggestion
    /// the `Event`'s [`dispatch` method]. For example, this happens internally
```
here --- the `tokio-trace` macros aren't the _only_ case where `Event::dispatch` will be called.

**Comment by hawkw** [comment]
@MSleepyPanda This is great, thanks again for your contribution! I'll merge this pending CI (though I doubt a docs change will break anything).

**Comment by MSleepyPanda** [comment]
Thank you for your time and patience! :) It's a blast having well reasoned and thorough feedback, even "just" for documentation

**Comment by hawkw** [comment]
@MSleepyPanda my pleasure! It's important to me that the docs are as clear as possible. I'm glad you appreciated my feedback and I'm sorry if it was overly nitpicky.
